### DISCUSSION TITLE
ampache http and 2 step authentication

### DISCUSSION THREAD

**Discussion by cnic023** [ampache/ampache#2998] [Q&A]
I'm trying to get 2 questions into 1 here.
1.   I'm putting ampache behind a reverse proxy. The letsencrypt certificate is provided by the proxy server and then I want the subsequent communications between the proxy and ampache (on a separate box) to be by http because it is on a private LAN. Do I have to change any of the configurations in the ampache folder? Or is it just a matter of ensuring that the virtual server (apache2) on the ampache box is set up properly?
2.   Is it possible to get an extra level of security for an admin login to ampache? For example, with yubikey or similar? Or is this something to suggest for future development?
Cheers
Chris

**Comment by lachlan-00** [comment]
1. you can set up http or https proxy between ampache and  your proxy.

(Usually) You need to configure your external port and host in the config file
![image](https://user-images.githubusercontent.com/1305249/131467807-cc5cf1f6-913f-4a66-9ff2-1f72a67665a2.png)

this tells ampache that you're using 443 https and the external url of the server.

**Comment by lachlan-00** [comment]
2 is only possible if you write that yourself at the moment. unless openid or one of the current auth methods provide that

**Comment by cnic023** [comment]
Hi Lachlan
Thanks for your advice. I have now (I think) solved the problem. I changed the http_host in the ampache config from ampache.example.com to the IP address of the ampache box on the LAN. Ampache now works if I login from within the LAN using http://IP:Port
However, I can't log in from the outside through the proxy server. All I get is an 'unpretty' login box that leads to a 404 error. I think the next step is to look again at my virtual host setup on the proxy server but I guess this is a Apache2 problem and has nothing to do with this discussion group. Do you agree or should I do a Q&A on github? I suppose there will be people trying to get ampache to work behind a reverse proxy and they may have ideas on the correct proxy config and could learn something from my experience?
Chris

**Reply by lachlan-00** [reply]
http host is the external url that your proxy is using
so if i proxied the demo server it would use
```
http_host =  "demo.ampache.dev"
```

local_web_path is proabbly where you should put the internal ip/url e.g. `http://192.168.1.3:9090`

**Comment by cnic023** [comment]
Hi Lachlan
Thanks for this advice. I changed the http_host and the local_web_path but no change in result. The explanation in the config file suggests that the local_web_path is only needed where the proxy and the ampache install are on the save server. In my case they are on separate servers.
I also put the IP etc of the proxy in the proxy sections toward the end of the config but I think these are designed for a forward proxy. Anyway, no change.
Chris

**Comment by stebe** [comment]
In my case with a reverse proxy, I set the local web path to the external https address and left the rest to the default values. Might work for you if you are using a split DNS.

**Comment by cnic023** [comment]
Thanks for that. I had already decided to change to the default settings. I then set the local web path to the external URL. However, on all combinations I still get this
***@***.***

**Reply by cnic023** [reply]
Can't seem to get this working. Has anyone managed to get a reverse proxy working between ampache and the internet where the reverse proxy is on a separate box to the ampache server but on the same private (192.168.0.0) subnet? I do have a split DNA in that I maintain a hosts file on the relevant servers (is that what a split DNS means?). My setup goes: client -> Internet -> ISP (static IP) Port Forward 443 to Proxy (192.168.0.xx:443) -> ampache box (192.168.0.yy:8080).  Caveat: I'm a newb at this level of networking.

**Reply by cnic023** [reply]
Just wondering if I have to change anything in the ampachedb? It is presently listening on port 80. Should this change to 8080?

**Reply by stebe** [reply]
You should be listening on port 443 if you are using https and you need to have the host files set on all the servers and clients to point to the reverse proxy IP address. It would be simpler to set up a DNS override on your router rather than maintain all those host files.

**Reply by cnic023** [reply]
It is set up so the proxy authenticates (listening on port 443) and subsequent messages to ampache server are http. To check my proxy config, I have set up a second virtual host on the proxy server in the same way i.e. client -> Internet -> proxy -> test server (rather than to ampache server). I get seamless connection to test index.html on test server. I'm thinking the ampache config is wrong because I am still getting:
<img width="271" alt="Capture" src="https://user-images.githubusercontent.com/71999448/132777778-525ba6c5-90c8-4770-89ef-a60b1b2a8e6a.PNG">

**Reply by cnic023** [reply]
This is part of my ampache.cfg.php
;#########################################################
; Path Vars                                              #
;#########################################################

; The public http host of your server.
; If not set, retrieved automatically from client request.
; This setting is required for WebSocket server
; DEFAULT: none
http_host = "ampacheserver.example.com"

; The public http port of your server.
; If not set, retrieved automatically from client request.
; DEFAULT: none
http_port = 8080

; The public path to your Ampache install
; Do not put a trailing / on this path
; For example if your site is located at http://localhost
; than you do not need to enter anything for the web_path
; if it is located at http://localhost/music you need to
; set web_path to /music
; DEFAULT: none
;web_path = /ampache

; The local http url of your server.
; This is used to access the server from within the
; same host where ampache is running.
; For example, if the ampache server is not
; directly accessed via the public domain but via a reverse
; proxy, local_web_path would need to be changed
; to a localhost URL.
; If not set, retrieved automatically from server information.
; DEFAULT: none
local_web_path = "http://ampacheserver.example.com:8080"

**Reply by cnic023** [reply]
Still no progress. I've used the exact same proxy and server setup for an installation of Nextcloud and that works fine (similar to ampache in that it uses LAMP). Is there some way to get a more detailed documentation for ampache.cfg.php?
